alpenhorn/Task.py

```diff
@@ -0,0 +1,1021 @@
+""" Classes to run each task in alpenhorn from a queue.
+"""
+
+import datetime as dt
+import logging
+import os
+import re
+import time
+import queue
+import threading
+
+import peewee as pw
+from peewee import fn
+
+from . import acquisition as ac
+from . import archive as ar
+from . import config, db
+from . import update
+from . import storage as st
+from . import util
+
+log = logging.getLogger(__name__)
+
+# Parameters.
+max_time_per_node_operation = 300  # Don't let node operations hog time.
+
+
+class TaskQueue:
+    def add_task(self, task):
+        """Add task to queue."""
+        if self.queue.full():
+            log.warning("Task queue is full ({:d} tasks)".format(self.queue.qsize()))
+        self.queue.put(task)
+
+    def run_tasks(self):
+        """Loop and run tasks from queue."""
+        while True:
+            try:
+                task = self.queue.get()
+            except queue.Empty:
+                pass
+            else:
+                task.run()
+                self.queue.task_done()
+
+    def __init__(self, max_size, num_threads):
+        """Create queue and setup thread pool."""
+        self.queue = queue.Queue(maxsize=max_size)
+
+        threads = []
+        for t in range(num_threads):
+            threads.append(threading.Thread(target=self.run_tasks, daemon=True))
+            threads[t].start()
+
+
+class Task:
+    def __init__(self, node):
+        self.node = node
+
+    def run(self):
+        print("{} run()...".format(type(self).__name__))
```

Comment:
Instead of being a print statement, this should probably be a `raise NotImplementedError`